File: backend/controllers/taskController.js
/**
 * taskController.js
 * -----------------
 * Handles all request/response logic for task CRUD.
 * Business rules live here; raw storage ops stay in taskStore.
 *
 * Task schema:
 *   id          — random 10-char alphanumeric string
 *   title       — string (trimmed, non-empty)
 *   isCompleted — boolean (default: false)
 *   date        — ISO date string of creation time
 */

const store = require('../storage/taskStore');

// ─── Utility ──────────────────────────────────────────────────────────────────

/**
 * Generate a random 10-character alphanumeric ID.
 * Charset: a-z A-Z 0-9  (62 chars → ~3.5 quadrillion combinations)
 */
const generateId = () => {
  const chars = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
  return Array.from({ length: 10 }, () =>
    chars[Math.floor(Math.random() * chars.length)]
  ).join('');
};

/** Build a fresh task object from a title string. */
const buildTask = (title) => ({
  id:          generateId(),
  title:       title.trim(),
  isCompleted: false,
  date:        new Date().toISOString(),
});

// ─── CREATE ───────────────────────────────────────────────────────────────────

/**
 * POST /api/tasks
 * Body: { "titles": ["Task A", "Task B", ...] }
 *
 * Creates one task per title string. Returns all created tasks.
 */
const createTasks = (req, res) => {
  const { titles } = req.body;

  // Validate: titles must be a non-empty array of strings
  if (!Array.isArray(titles) || titles.length === 0) {
    return res.status(400).json({
      success: false,
      error: '"titles" must be a non-empty array of strings.',
    });
  }

  const invalid = titles.filter((t) => typeof t !== 'string' || t.trim() === '');
  if (invalid.length > 0) {
    return res.status(400).json({
      success: false,
      error: 'All entries in "titles" must be non-empty strings.',
      invalid_entries: invalid,
    });
  }

  const newTasks = titles.map(buildTask);
  const inserted = store.insertMany(newTasks);

  return res.status(201).json({
    success: true,
    message: `${inserted.length} task(s) created.`,
    count:   inserted.length,
    tasks:   inserted,
  });
};

// ─── READ ALL ─────────────────────────────────────────────────────────────────

/**
 * GET /api/tasks
 * Query params (all optional):
 *   ?completed=true|false   filter by completion status
 *   ?search=keyword         filter titles containing keyword (case-insensitive)
 *
 * Returns all tasks, newest first.
 */
const getAllTasks = (req, res) => {
  let tasks = store.getAll();

  // Filter: isCompleted
  if (req.query.completed !== undefined) {
    const want = req.query.completed === 'true';
    tasks = tasks.filter((t) => t.isCompleted === want);
  }

  // Filter: title search
  if (req.query.search) {
    const kw = req.query.search.toLowerCase();
    tasks = tasks.filter((t) => t.title.toLowerCase().includes(kw));
  }

  // Sort: newest first (by date desc)
  tasks = tasks.slice().sort((a, b) => new Date(b.date) - new Date(a.date));

  return res.json({
    success: true,
    count:   tasks.length,
    tasks,
  });
};

// ─── READ ONE ─────────────────────────────────────────────────────────────────

/**
 * GET /api/tasks/:id
 * Returns a single task by its id.
 */
const getTaskById = (req, res) => {
  const task = store.getById(req.params.id);

  if (!task) {
    return res.status(404).json({
      success: false,
      error:   `Task with id "${req.params.id}" not found.`,
    });
  }

  return res.json({ success: true, task });
};

// ─── UPDATE ───────────────────────────────────────────────────────────────────

/**
 * PATCH /api/tasks/:id
 * Body: { "title": "New title", "isCompleted": true }
 *
 * Partially updates a task. Only "title" and "isCompleted" are mutable.
 * id and date are immutable and silently ignored if sent.
 */
const updateTask = (req, res) => {
  const { title, isCompleted } = req.body;

  // At least one valid field required
  if (title === undefined && isCompleted === undefined) {
    return res.status(400).json({
      success: false,
      error:   'Provide at least one field to update: "title" or "isCompleted".',
    });
  }

  // Validate types
  if (title !== undefined && (typeof title !== 'string' || title.trim() === '')) {
    return res.status(400).json({ success: false, error: '"title" must be a non-empty string.' });
  }

  if (isCompleted !== undefined && typeof isCompleted !== 'boolean') {
    return res.status(400).json({ success: false, error: '"isCompleted" must be a boolean.' });
  }

  const fields = {};
  if (title       !== undefined) fields.title       = title.trim();
  if (isCompleted !== undefined) fields.isCompleted = isCompleted;

  const updated = store.updateById(req.params.id, fields);

  if (!updated) {
    return res.status(404).json({
      success: false,
      error:   `Task with id "${req.params.id}" not found.`,
    });
  }

  return res.json({ success: true, message: 'Task updated.', task: updated });
};

// ─── DELETE ONE ───────────────────────────────────────────────────────────────

/**
 * DELETE /api/tasks/:id
 * Removes a single task. Returns the deleted task for confirmation.
 */
const deleteTask = (req, res) => {
  const deleted = store.deleteById(req.params.id);

  if (!deleted) {
    return res.status(404).json({
      success: false,
      error:   `Task with id "${req.params.id}" not found.`,
    });
  }

  return res.json({ success: true, message: 'Task deleted.', task: deleted });
};

// ─── DELETE ALL ───────────────────────────────────────────────────────────────

/**
 * DELETE /api/tasks
 * Clears all tasks from the store. Use with care.
 */
const deleteAllTasks = (req, res) => {
  const count = store.clearAll();
  return res.json({ success: true, message: `${count} task(s) deleted.`, count });
};

module.exports = {
  createTasks,
  getAllTasks,
  getTaskById,
  updateTask,
  deleteTask,
  deleteAllTasks,
};

File: backend/allFastAPI/controllers/taskController.py
import string
import random
from datetime import datetime, UTC
from typing import List, Optional
from fastapi import HTTPException, status
from pydantic import BaseModel, Field

# Assuming your store logic is in a file named task_store.py
# from storage import task_store as store

# --- Models (Schemas) ---

class Task(BaseModel):
    id: str
    title: str
    isCompleted: bool = False
    date: str

class CreateTasksRequest(BaseModel):
    # This automatically handles the "non-empty array of strings" validation
    titles: List[str] = Field(..., min_items=1)

class UpdateTaskRequest(BaseModel):
    title: Optional[str] = None
    isCompleted: Optional[bool] = None

# --- Utility ---

def generate_id(length=10):
    chars = string.ascii_letters + string.digits
    return ''.join(random.choice(chars) for _ in range(length))

def build_task(title: str) -> dict:
    return {
        "id": generate_id(),
        "title": title.strip(),
        "isCompleted": False,
        "date": datetime.now(UTC).isoformat()
    }

# --- CREATE ---

async def create_tasks(payload: CreateTasksRequest):
    # Pydantic already validated that titles is a list of strings
    if any(not t.strip() for t in payload.titles):
        raise HTTPException(
            status_code=400, 
            detail="All entries in 'titles' must be non-empty strings."
        )

    new_tasks = [build_task(t) for t in payload.titles]
    # inserted = store.insert_many(new_tasks)
    inserted = new_tasks # Placeholder for store logic

    return {
        "success": True,
        "message": f"{len(inserted)} task(s) created.",
        "count": len(inserted),
        "tasks": inserted
    }

# --- READ ALL ---

async def get_all_tasks(completed: Optional[bool] = None, search: Optional[str] = None):
    # tasks = store.get_all()
    tasks = [] # Placeholder

    if completed is not None:
        tasks = [t for t in tasks if t['isCompleted'] == completed]

    if search:
        kw = search.lower()
        tasks = [t for t in tasks if kw in t['title'].lower()]

    # Sort: newest first
    tasks.sort(key=lambda x: x['date'], reverse=True)

    return {"success": True, "count": len(tasks), "tasks": tasks}

# --- READ ONE ---

async def get_task_by_id(task_id: str):
    # task = store.get_by_id(task_id)
    task = None # Placeholder
    
    if not task:
        raise HTTPException(status_code=404, detail=f"Task with id '{task_id}' not found.")
    
    return {"success": True, "task": task}

# --- UPDATE ---

async def update_task(task_id: str, payload: UpdateTaskRequest):
    if payload.title is None and payload.isCompleted is None:
        raise HTTPException(status_code=400, detail='Provide "title" or "isCompleted".')

    if payload.title is not None and not payload.title.strip():
        raise HTTPException(status_code=400, detail='"title" cannot be empty.')

    fields = payload.model_dump(exclude_unset=True)
    if 'title' in fields:
        fields['title'] = fields['title'].strip()

    # updated = store.update_by_id(task_id, fields)
    updated = None # Placeholder

    if not updated:
        raise HTTPException(status_code=404, detail=f"Task with id '{task_id}' not found.")

    return {"success": True, "message": "Task updated.", "task": updated}

# --- DELETE ---

async def delete_task(task_id: str):
    # deleted = store.delete_by_id(task_id)
    deleted = None # Placeholder
    
    if not deleted:
        raise HTTPException(status_code=404, detail=f"Task with id '{task_id}' not found.")
    
    return {"success": True, "message": "Task deleted.", "task": deleted}

async def delete_all_tasks():
    # count = store.clear_all()
    count = 0
    return {"success": True, "message": f"{count} task(s) deleted.", "count": count}
